multinet/api/models/workspace.py

```diff
@@ -1,124 +1,116 @@
 from __future__ import annotations
 
-from typing import Type
+from typing import Type, Union
 from uuid import uuid4
 
 from arango.database import StandardDatabase
 from django.contrib.auth.models import User
-from django.db.models import BooleanField, CharField
+from django.db import models
 from django.db.models.signals import post_delete, pre_save
 from django.dispatch import receiver
 from django_extensions.db.models import TimeStampedModel
-from guardian.shortcuts import assign_perm, get_user_perms, get_users_with_perms, remove_perm
 
 from multinet.api.utils.arango import ensure_db_created, ensure_db_deleted, get_or_create_db
-from multinet.api.utils.workspace_permissions import WorkspacePermission
 
 
 def create_default_arango_db_name():
     # Arango db names must start with a letter
     return f'w-{uuid4().hex}'
 
 
-class Workspace(TimeStampedModel):
-    name = CharField(max_length=300, unique=True)
-    public = BooleanField(default=False)
+class WorkspaceRoleChoice(models.IntegerChoices):
+    READER = 1
+    WRITER = 2
+    MAINTAINER = 3
 
-    # Max length of 34, since uuid hexes are 32, + 2 chars on the front
-    arango_db_name = CharField(max_length=34, unique=True, default=create_default_arango_db_name)
+
+class WorkspaceRole(TimeStampedModel):
+    workspace = models.ForeignKey('api.Workspace', on_delete=models.CASCADE)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    role = models.PositiveSmallIntegerField(choices=WorkspaceRoleChoice.choices)
 
     class Meta:
-        ordering = ['id']
-        permissions = [
-            (WorkspacePermission.owner.name, 'Owns the workspace'),
-            (WorkspacePermission.maintainer.name, 'Grants roles except owner on the workspace'),
-            (WorkspacePermission.reader.name, 'Creates and deletes data on the workspace'),
-            (WorkspacePermission.writer.name, 'Views data on the workspace'),
+        constraints = [
+            models.UniqueConstraint(
+                fields=['workspace', 'user'], name='unique_workspace_permission'
+            ),
         ]
 
-    @property
-    def owner(self):
-        """
-        Return the workspace owner.
 
-        Django-guardian allows for potentially more than one user with
-        the 'owner' permission, so we must take care to limit the number of owners ourselves by
-        using this property and Workspace.set_owner to handle assigning owners.
-        """
-        return get_users_with_perms(
-            self, only_with_perms_in=[WorkspacePermission.owner.name]
-        ).first()
+class Workspace(TimeStampedModel):
+    name = models.CharField(max_length=300, unique=True)
+    public = models.BooleanField(default=False)
+    # What should happen to a workspace when the owner is deleted?
+    owner = models.ForeignKey(User, on_delete=models.PROTECT)
```

Comment:
How are users deleted? Can they delete themselves or is that something that can only be done by a Django superuser? If users can delete themselves then I think `CASCADE` is fine, but maybe they should have the option to reassign at the time of deletion. That's more of a frontend/workflow concern though.

Actually now that I'm thinking about it, `CASCADE` in general probably fine, but there is a possibility that people lose access to a workspace they care about/use frequently if that workspace owner is deleted.

Reply:
Deleting a user should delete all their workspaces. This is because of a couple of factors:
1. We need to have a general stance that users can easily remove themselves and their data from the system, a la the EU's right to be forgotten.
2. Owners have the ability to transfer ownership to someone else, so they can do that in preparation for deleting their accounts.

What this means is, the client must have UI warning a user that their workspaces will be destroyed when they delete their account, and give them a chance to avoid that outcome. We should do so by making them type a random phrase to confirm deletion (like we do for tables and networks).

It also means that in the docs for the API, we need to make sure users understand how dangerous the `DELETE /users/me` endpoint is. We can't enforce a confirmation mechanism on that endpoint, by design.

I prefer this approach for the reasons above; additionally, we would have to decide on and implement quite a bit of logic if we _don't_ delete workspaces on user deletion. (i.e., who would own the workspaces? do they become ownerless? what are the semantics of an ownerless workspace? etc.)

Reply:
> How are users deleted? Can they delete themselves or is that something that can only be done by a Django superuser? If users can delete themselves then I think `CASCADE` is fine, but maybe they should have the option to reassign at the time of deletion. That's more of a frontend/workflow concern though.

I think we need to discuss this and make a decision. In many systems there's a "request deletion" action, and then a human will carry that out. That's what Facebook does, but I believe that's mainly because they want to gaslight you into not deleting your account before you make a final decision. We have no such designs in Multinet, so I think self-deletion would be fine. But as I mentioned above, I'd want to heavily protect the user from doing this lightly, since it's irreversible. (Or rather, it would be costly to make it reversible, and I don't have any desire to make it so.)

> 
> Actually now that I'm thinking about it, `CASCADE` in general probably fine, but there is a possibility that people lose access to a workspace they care about/use frequently if that workspace owner is deleted.

Can you say more about the "possibility"? I would describe it more strongly--it's an "eventuality" if they self-delete, right?

Reply:
Ok, in light of point (1) I'll change this to `CASCADE`. I think its important to make sure the user is informed of the side effects of their deletion, and I agree that they should be given the chance to avoid that outcome. I know that deletion of tables/networks prompts a random phrase, but I'm unaware if this is already in place for users. If not, it or something similar should be.

Reply:
> Can you say more about the "possibility"? I would describe it more strongly--it's an "eventuality" if they self-delete, right?

My concern is a scenario where, say User A owns some workspace that User B frequently uses for whatever reason. Maybe the workspace is public, so A and B have no real-world knowledge of each other. If User A deletes themself, then User B no longer has access to that workspace, which might not have a huge impact, but it would feel bad to be User B in that situation. 

Thinking about this situation, its effectively the same as if User A just deletes that workspace, which is a reasonable workflow. There is a difference though. Maybe User A deleting the workspace is could be communicating a message like "the information contained within this workspace is not useful/incorrect/outdated/etc, and should be removed." User A deleting themself doesn't necessarily have this connotation. 

Maybe I'm overthinking this situation, or I'm still lacking some context for real world use cases of multinet.

Reply:
Ah I see. Yes, that would be unfortunate for User B but I don't see what can really be done about it "in-band". This is the same as a YouTube video disappearing, or countless other "shared repository" examples, and as such I think we don't really have to account for this type of scenario.

The generally envisioned usage of Multinet is that workspace owners are collecting data there in order to conduct their own research, etc. The sharing features are meant to allow for others to collaborate on that data, but there's no strong sense of entitlement built in for the people to whom a workspace is shared.

_However_, one easy way we could solve these problems is by allowing multiple owners on a workspace. Then, when a user deletes themselves, a workspace with remaining owners wouldn't be deleted. But I also don't think it's worthwhile for us to change our model of ownership and permissions just to make this specific scenario easier on hypothetical users.

Reply:
I agree with @waxlamp. I think for now, we shouldn't worry about this scenario. If it does become a problem, multiple owners would make sense, and could be easily incorporated by changing the workspace `owner` field from a `ForeignKey` to a `ManyToManyField`.

Reply:
This all sounds good to me. Thanks guys.

Reply:
Reminder to the thread-opener to resolve this if/when you are satisfied with the discussion 😸